// usehooks-ts/dist/cjs/useStep/useStep.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useStep = void 0;
var react_1 = require("react");
/**
 * Custom hook for managing and navigating between steps in a multi-step process.
 * @param {number} maxStep - The maximum step in the process.
 * @returns {[number, Helpers]} An tuple containing the current step and helper functions for navigating steps.
 * @see [Documentation](https://usehooks-ts.com/react-hook/use-step)
 * @example
 * const [currentStep, { goToNextStep, goToPrevStep, reset, canGoToNextStep, canGoToPrevStep, setStep }] = useStep(3);
 * // Access and use the current step and provided helper functions.
 */
function useStep(maxStep) {
    var _a = (0, react_1.useState)(1), currentStep = _a[0], setCurrentStep = _a[1];
    var canGoToNextStep = (0, react_1.useMemo)(function () { return currentStep + 1 <= maxStep; }, [currentStep, maxStep]);
    var canGoToPrevStep = (0, react_1.useMemo)(function () { return currentStep - 1 >= 1; }, [currentStep]);
    var setStep = (0, react_1.useCallback)(function (step) {
        // Allow value to be a function so we have the same API as useState
        var newStep = step instanceof Function ? step(currentStep) : step;
        if (newStep >= 1 && newStep <= maxStep) {
            setCurrentStep(newStep);
            return;
        }
        throw new Error('Step not valid');
    }, [maxStep, currentStep]);
    var goToNextStep = (0, react_1.useCallback)(function () {
        if (canGoToNextStep) {
            setCurrentStep(function (step) { return step + 1; });
        }
    }, [canGoToNextStep]);
    var goToPrevStep = (0, react_1.useCallback)(function () {
        if (canGoToPrevStep) {
            setCurrentStep(function (step) { return step - 1; });
        }
    }, [canGoToPrevStep]);
    var reset = (0, react_1.useCallback)(function () {
        setCurrentStep(1);
    }, []);
    return [
        currentStep,
        {
            goToNextStep: goToNextStep,
            goToPrevStep: goToPrevStep,
            canGoToNextStep: canGoToNextStep,
            canGoToPrevStep: canGoToPrevStep,
            setStep: setStep,
            reset: reset,
        },
    ];
}
exports.useStep = useStep;
